src/riak_core_info_service_process.erl

```diff
@@ -81,20 +90,55 @@ terminate_not_implemented(Request, State) ->
 handle_invoke_message(ProviderParams, HandlerContext,
                       #state{service_provider = Provider,
                              response_handler = ResponseHandler } = State) ->
-    Result = apply_callback(Provider, ProviderParams),
+    %% At this layer we don't care much whether a callback succeeded
+    {_, Result} = apply_callback(Provider, ProviderParams),
+
     Reply = {Result, ProviderParams, HandlerContext},
-    apply_callback(ResponseHandler, [Reply]),
-    {noreply, State}.
+    handle_consumer_response(apply_callback(ResponseHandler, [Reply]), response, State).
 
 handle_shutdown_message(#state{shutdown = Shutdown}=State) ->
-    apply_callback(Shutdown, [self()]),
+    _ = apply_callback(Shutdown, [self()]),
     {stop, normal, State}.
 
-apply_callback(undefined, _CallSpecificArgs) ->
-    ok;
+%% The original response to the callback will be wrapped in an `ok' or
+%% `exit' tuple, primarily so that the service process can terminate
+%% if something goes wrong during registration.
+-spec apply_callback(callback(), list(term())) -> {ok, term()} |
+                                                  {error, term()}.
 apply_callback({Mod, Fun, StaticArgs}, CallSpecificArgs) ->
     %% Invoke the callback, passing static + call-specific args for this call
-    erlang:apply(Mod, Fun, StaticArgs ++ CallSpecificArgs).
+    Args = StaticArgs ++ CallSpecificArgs,
+    _ = case (catch erlang:apply(Mod, Fun, Args)) of
```

Comment:
with `catch`, a `throw` in `erlang:apply(Mod, Fun, Args)` matches `Result` and is indistinguishable from a return value. If this is not the intention, a `try-catch` should be used instead.